unity/JetBrains.Rider.Unity.Editor/EditorPlugin/AssetPostprocessors/CsprojAssetPostprocessor.cs

```diff
@@ -332,54 +333,89 @@ private static void ApplyCustomReference(string name, XElement projectContentEle
     // Set appropriate version
     private static void FixTargetFrameworkVersion(XElement projectElement, XNamespace xmlns)
     {
-        var targetFrameworkVersion = projectElement.Elements(xmlns + "PropertyGroup")
-          .Elements(xmlns + "TargetFrameworkVersion")
-          .FirstOrDefault(); // Processing csproj files, which are not Unity-generated JetBrains/Unity3dRider#56
-        if (targetFrameworkVersion == null)
-          return;
-
-      if (UnityUtils.ScriptingRuntime > 0)
+      SetOrUpdateProperty(projectElement, xmlns, "TargetFrameworkVersion", s =>
         {
-          var version = PluginSettings.TargetFrameworkVersion;
-          if (!PluginSettings.OverrideTargetFrameworkVersion)
+          string version = string.Empty;
+          try
           {
+            version = s.Substring(1);
+            // for windows try to use installed dotnet framework
             if (PluginSettings.SystemInfoRiderPlugin.operatingSystemFamily == OperatingSystemFamilyRider.Windows)
             {
               var versions = PluginSettings.GetInstalledNetFrameworks();
               if (versions.Any())
               {
-                version = versions.OrderBy(v => new Version(v)).Last();
+                var versionOrderedList = versions.OrderBy(v1 => new Version(v1));
+                var foundVersion = UnityUtils.ScriptingRuntime > 0
+                  ? versionOrderedList.Last()
+                  : versionOrderedList.First();
+                // roslyn compiler requires dotnet 4.7.1, which may not be present
+                var fvIsParsed = decimal.TryParse(foundVersion, NumberStyles.Number, CultureInfo.InvariantCulture, out var fv);
+                var vIsParsed = decimal.TryParse(version, NumberStyles.Number, CultureInfo.InvariantCulture, out var v);
+                if (fvIsParsed && vIsParsed && fv > v)
```

Comment:
Why parse as decimals? `Version` allows comparison

Reply:
These are strings. And dotnet 3.5 doesn't have Version.TryParse

Reply:
Just parse as `Version` but inside a try/catch? Parsing as decimal will fail with 4.6.1 and 4.6.2

Reply:
switched to Version